### ISSUE TITLE
[BUG] - Block producer fails to adopt block when trying to switch fork immediately after forging

### ISSUE THREAD

**Issue by renesecur** [IntersectMBO/cardano-node#5454]
Has anyone ever seen the following on a BP?

-> 2023-08-26T20:03:00.00Z TraceStartLeadershipCheck
-> 2023-08-26T20:03:00.03Z TraceNodeIsLeader
-> 2023-08-26T20:03:00.08Z TraceForgedBlock
-> 2023-08-26T20:03:00.09Z TraceAddBlockEvent.AddedToCurrentChain (not the block just forged)
-> 2023-08-26T20:03:00.09Z TraceAddBlockEvent.AddedToCurrentChain (not the block just forged)
-> 2023-08-26T20:03:00.09Z TraceAddBlockEvent.TrySwitchToAFork
-> 2023-08-26T20:03:00.13Z TraceDidntAdoptBlock (Error) !

What it looks like to me is something fishy happens between Forging and adopting and that led to an error. The forged block is not logged as a valid candidate and I see two statements adding the same block, which is a blockhash that never made it on-chain either. This block is identified by b8996a97f7115ee0efd46ba10e08b03d85b13531b81e8762f982d03a0f2596dc@101513862 and was forged 27 (!) seconds before ours

Our block was scheduled at slot 101513889 which is square in the middle of two other blocks that made it on-chain fine so it doesn't look like anything height-battle related. Between block 9210211 on slot 101513859 and block 9210212 on slot 101513963

The normal statements are 

-> TraceStartLeadershipCheck
-> TraceNodeIsLeader
-> TraceForgedBlock
-> TraceAddBlockEvent.AddBlockValidation.ValidCandidate
-> TraceAddBlockEvent.AddedToCurrentChain
-> TraceAdoptedBlock

From what I can tell it usually take 30 ms between forging the block and logging the valid candidate. In our case some other block was added to current chain 10 ms after forging the block so the node was supposedly in the process of validating the just forged block.

Is it worth logging a report for this in cardano-node or is it a known issue?

Pretty odd to see this...

We are running on `8.1.2:d2d90` in Docker on an ARM instance AWS type r6g.xlarge. Built with ghc 8.10.7 and cabal 3.8.1.0.

**Comment by lead-pool** [comment]
I believe the same thing has occurred to me last epoch - the cause was due to a pool having a time sync issue minting a block too early causing a fork - my block was then minted on the fork as we believe i was in their peer group, as my block was on the fork it was obviously not adopted.

@gufmar helped analyse (thanks again!)

Just for the sake of documenting this issue a bit further and adding another example

Epoch: 433
Block with height battle: 9219198
Slot where LEAD expected to mint a block: 15908

**Comment by renesecur** [comment]
Yes you may be correct. The block hash in our logs and the timings correspond to an orphaned block detected by pooltool https://pooltool.io/realtime/9210212

I still find 2 things strange: why and how the block is added to the chain 10ms after forging our own block. And why the node fails to adopt the forged block after switching forks. This should be locally resolvable no?
